 0016-capi-upgrades/0016-capi-upgrades.md

```diff
@@ -0,0 +1,49 @@
+# RFC 016 - CLUSTER API UPGRADES
+
+## Reasoning
+
+As part of the Cluster API (CAPI) Hive Sprint, the `upgrades hive` decided to create a document to gather all actions need to be done in order to upgrade a workload cluster successfully as well as bring the functionality we have already built to trigger the upgrades.
+
+## Scope
+
+This document covers only the action to upgrade the workload cluster without solving problems related to the versioning of the different CAPI controllers and/or deciding which controller reconciles a cluster.
+
+At the same time all the automation on top of the upgrades, like freeze periods or maintainance windows, are out of the scope here ([include in RFC 0015](/0015-automatic-cluster--upgrades)).
+
+Furthermore, a workload cluster can be upgraded in different ways:
+
+- Apps installed in the cluster are upgraded (out of scope here). We have other non-CAPI components that already take care of that, like app-operator
+
+- Cluster is reconciled by a newer CAPI controller. We may want to upgrade the Cluster to a newer CAPI controller release, which may change the infrastructure created to support the cluster (network, load balancers, etc). We do this update just changing the CAPI label on the Cluster CR, and we don't need a new machine image to be created/referenced by the CR's.
+
+- Machine image used by control plane or workers is updated with new k8s version or new OS version. We may want to update the k8s version or the OS version running in the machines. In both cases, we need a new machine image using the updated components, and reference that image from the CR's.
+
+## Context
+
+The current approach to upgrade a workload cluster in CAPI world involves a series of steps described below:
+
+- Ensure there is a `machine image` with the correct target Kubernetes version (contains kubelet and kubeadm). 
+- Upgrade the control plane machines:
+  - Generate a new infrastructure `MachineTemplate` resource (with new image ID if needed) and create it in the Management API.
+  - Set the new Kubernetes version in `KubeadmControlPlane` resource.
+  - Change the infrastructure `MachineTemplate` on the `KubeadmControlPlane` to the new one and apply the new version of `KubeadmControlPlane`.
+- Upgrade machine pools. As default, we walk through all existing machine pools one by one:
+  - Make sure there are no MachinePools being already upgraded.
+  - Modify the infrastructure `MachinePool` resource (with new image ID) if needed.
+  - Set the new Kubernetes version in `MachinePool` resource.
+
+Also our operators watch for changes on the Cluster CR's labels to trigger an upgrade. That functionality should be kept in the new operator we build.
```

Comment:
```suggestion
The upgrade operator watches for changes on the Cluster CR's labels to trigger an upgrade. In case the release does not bring an OS or k8s change the upgrade operator will not trigger the process.
```